pkg/commonvpc/parser.go

```diff
@@ -0,0 +1,307 @@
+	}
+}
+
+func GetRegionByName(regionName string, regionToStructMap map[string]*Region) *Region {
+	regionPointer, ok := regionToStructMap[regionName]
+
+	if !ok {
+		regionToStructMap[regionName] = &Region{Name: regionName}
+		return regionToStructMap[regionName]
+	}
+	return regionPointer
+}
+
+func NewVPC(name, uid, region string, zonesToAP map[string][]string, regionToStructMap map[string]*Region) (
+	vpcNodeSet *VPC, err error) {
+	var regionPointer *Region
+	if regionToStructMap != nil {
+		regionPointer = GetRegionByName(region, regionToStructMap)
+	}
+	vpcNodeSet = &VPC{
+		VPCResource: vpcmodel.VPCResource{
+			ResourceName: name,
+			ResourceUID:  uid,
+			ResourceType: ResourceTypeVPC,
+			Region:       region,
+		},
+		Zones:     map[string]*Zone{},
+		VPCregion: regionPointer,
+		VPCnodes:  []vpcmodel.Node{},
+	}
+	for zoneName, zoneCidrsList := range zonesToAP {
+		vpcNodeSet.AddressPrefixesList = append(vpcNodeSet.AddressPrefixesList, zoneCidrsList...)
+		zoneIPBlock, err := ipblock.FromCidrList(zoneCidrsList)
+		if err != nil {
+			return nil, err
+		}
+		vpcNodeSet.Zones[zoneName] = &Zone{Name: zoneName,
+			Vpc:     vpcNodeSet,
+			Cidrs:   zoneCidrsList,
+			IPblock: zoneIPBlock}
+	}
+
+	vpcNodeSet.AddressPrefixesIPBlock, err = ipblock.FromCidrList(vpcNodeSet.AddressPrefixesList)
+	if err != nil {
+		return nil, err
+	}
+	vpcNodeSet.VPCRef = vpcNodeSet
+	return vpcNodeSet, nil
+}
+
+func GetVPCconfigByUID(res *vpcmodel.MultipleVPCConfigs, uid string) (*vpcmodel.VPCConfig, error) {
+	vpcConfig, ok := res.Configs()[uid]
+	if !ok {
+		return nil, fmt.Errorf("missing VPC resource of uid %s", uid)
+	}
+	return vpcConfig, nil
+}
+
+func GetVPCObjectByUID(res *vpcmodel.MultipleVPCConfigs, uid string) (*VPC, error) {
+	vpcConfig, err := GetVPCconfigByUID(res, uid)
+	if err != nil {
+		return nil, err
+	}
+	vpc, ok := vpcConfig.VPC.(*VPC)
+	if !ok {
+		return nil, fmt.Errorf("VPC missing from config of VPCConfig with uid %s", uid)
+	}
+	return vpc, nil
+}
+
+func AddZone(zoneName, vpcUID string, res *vpcmodel.MultipleVPCConfigs) error {
+	vpc, err := GetVPCObjectByUID(res, vpcUID)
+	if err != nil {
+		return err
+	}
+	if _, ok := vpc.Zones[zoneName]; !ok {
+		vpc.Zones[zoneName] = &Zone{Name: zoneName, Vpc: vpc}
+	}
+	return nil
+}
+
+func NewSubnet(name, uid, zone, cidr string, vpc vpcmodel.VPCResourceIntf) (*Subnet, error) {
+	subnetNode := &Subnet{
+		VPCResource: vpcmodel.VPCResource{
+			ResourceName: name,
+			ResourceUID:  uid,
+			Zone:         zone,
+			ResourceType: ResourceTypeSubnet,
+			VPCRef:       vpc,
+			Region:       vpc.RegionName(),
+		},
+		Cidr: cidr,
+	}
+
+	cidrIPBlock, err := ipblock.FromCidr(subnetNode.Cidr)
+	if err != nil {
+		return nil, err
+	}
+	subnetNode.IPblock = cidrIPBlock
+	return subnetNode, nil
+}
+
+func PrintLineSection() {
+	logging.Debugf("-----------------------------------------")
+}
+
+func PrintSGRules(sg *SecurityGroup) {
+	numRules := sg.Analyzer.SgAnalyzer.GetNumberOfRules()
```

Comment:
You could use `GetRules()` for `SecurityGroupLayer` instead of the above  `GetNumberOfRules() `and `sg.Analyzer.SgAnalyzer.GetSGRule(i)` below